add login.gov provider

timothy-spencer opened the pull request:
## Description

This change does three things:
  * Adds the login.gov provider, which is basically an OIDC provider, but doesn't have renewal stuff in it, and it also requires more parameters for the authentication step.
  * Adds environment variables to all config options, so that it is easier to configure the proxy in environments like GCP's App Engine.
  * Updates the README with info about how to use the login.gov provider.

## Motivation and Context

We would like to be able to use the proxy with login.gov.

## How Has This Been Tested?

I wrote tests into logingov_test.go, and have been using that.
I also have set up a test application using the login.gov integration environment on my local host,
using the instructions in the README.md.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [X] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

timothy-spencer commented:
Hey, I don't understand those linter errors.  I'm happy to adjust stuff on my branch if you can help me figure them out.  :-)  I have tried doing a gofmt on the files, which results in no differences. 🤷‍♂️

ploxiln commented:
```
options.go:1::warning: file is not goimported (goimports)
providers/logingov.go:1::warning: file is not goimported (goimports)
providers/logingov_test.go:1::warning: file is not goimported (goimports)
```

These are not code-style warnings. This seems to happen only in the source files in which you import "github.com/dgrijalva/jwt-go". I'm not sure why ... but my first guess is that it wants you to give this import a name because the package name does not match the last component of the path. Like:

```go
import (
        jwt "github.com/dgrijalva/jwt-go"
)
```

Notice also that this codebase puts non-standard-library imports in a separate grouping that's an informal code-style thing.

timothy-spencer commented:
Perfect.  The separate grouping seemed to do the trick!  Thanks!

ploxiln commented:
Huh, I guess it _was_ a code-style thing :) I think "goimport" is like "gofmt" with even more rules for imports, so it was saying that "goimport" must not have been not run on these files. Good to know ...

JoelSpeed commented:
Is this ready for review?

timothy-spencer commented:
Yes, I believe it is ready for review.  Please feel free to give it the approval/feedback/etc that it deserves!  :-)

timothy-spencer commented:
Hey, I know this is a pretty big set of changes, but it would be great if it could get some review/approval love so that it won't get stale and hard to merge.  Just wanted to ping to make sure that this was still on your mind.  :-)  Thanks!

JoelSpeed commented on providers/logingov.go:
```diff
@@ -0,0 +1,216 @@
+	params.Set("response_type", "code")
+	params.Add("state", state)
+	params.Add("acr_values", p.AcrValues)
+	params.Add("nonce", p.Nonce)
```
What is the purpose of this nonce? It seems to only be used here? Is it strictly required?

timothy-spencer replied:
It is required:  https://developers.login.gov/oidc/#authorization , which refers to https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes  Without it and the acr_values param, login.gov will refuse to authorize.

timothy-spencer replied:
FYI, just plugged in a check to make sure that the nonce is proper too.

timothy-spencer replied:
One concern that I do have about the nonce is that it really ought to be done per-session, but the session state is created in OAuthCallback, not OAuthStart when it needs to be created/stored.  It seems like this would take a lot of restructuring to fix, unless you have any great ideas.

JoelSpeed replied:
I can't see a way to easily create the nonce on a per session basis at the moment, but I agree it is something we should probably do in the long term. Could you add a TODO somewhere to signify that we should do this?

Also, as far as I can see, we aren't currently using the nonce when creating the signed JWT, is that correct or have I missed something?

timothy-spencer replied:
To clarify, the nonce is supplied by us in the authorization step, and then it is given back in the signed JWT id_token that we get back from the login.gov IDP during the redeem step.  We do not supply or use the nonce in any other context.  I now have code that checks the signature and the nonce in the id_token, so we do everything we can to make sure that everything is proper.  All that remains is for us to figure out how to supply a unique nonce per session.

JoelSpeed commented on providers/logingov_test.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+func newLoginGovProvider() (l *LoginGovProvider, err error) {
+	keyData, err := ioutil.ReadFile("sample_key")
```
Is it possible to generate a key on the fly rather than using a stored key within the repository? I worry that having example keys checked in may lead someone to use it somewhere where they shouldn't be, WDYT?

timothy-spencer replied:
It probably is.  I tried to do that at one point and found it to be hard in some way that I cannot remember, but you are right.  It would be better that way, so I'll give it another try.  :-)

timothy-spencer commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Changes since v3.1.0
 
+- Added login.gov provider
```
Will do!

timothy-spencer commented:
Sorry, didn't mean to harass you!  Life is indeed busy.  :-)  I have updated everything that you requested and did a merge with master, updated CODEOWNERS, etc.  Let me know what else you need.  Thanks!

JoelSpeed commented on providers/logingov.go:
```diff
@@ -0,0 +1,275 @@
+		return
+	}
+
+	claims := &jwt.StandardClaims{
```
Should the Nonce be going into the claims here?

timothy-spencer replied:
No, it's only used in the id_token to verify that the id_token (https://developers.login.gov/oidc/#token-response) is one that was generated from the authorization call where we supplied the nonce originally (https://developers.login.gov/oidc/#authorization).

timothy-spencer commented:
By a TODO, did you mean in the code, or an issue in the repo, or something in the README?  I would think that an issue would be the best way to do this, but wanted to make sure in case you had something else in mind.  :-)

JoelSpeed commented:
I would tend to put a `TODO(@<gh_id>): ...` in the code and raise an issue to match it. Helps people reading the code notice the known issues IMO, either is sufficient on its own though

timothy-spencer commented:
Hey, anything else you need me to do here?  I actually have a few more changes that I am maintaining on another branch that I am using to be able to deploy this stuff in GCP that I'd love to clean up and PR, but it would be hard without all the rest of this stuff.  :-)  Love to get this merged in!

JoelSpeed commented:
@timothy-spencer Thanks for your work on this 😄

timothy-spencer commented:
Thank _you_ for taking the time to review/merge!
